smart-contracts/wasm-chain-integration/src/utils.rs

```diff
@@ -337,6 +337,60 @@ pub fn generate_contract_schema_v2(
     }))
 }
 
+/// Tries to generate schemas for parameters and return values of methods for a
+/// contract with a V3 schema.
+pub fn generate_contract_schema_v3(
+    module_bytes: &[u8],
+) -> ExecResult<schema::VersionedModuleSchema> {
+    let artifact = utils::instantiate::<ArtifactNamedImport, _>(&TestHost, module_bytes)?;
+
+    let mut contract_schemas = BTreeMap::new();
+
+    for name in artifact.export.keys() {
+        if let Some(rest) = name.as_ref().strip_prefix("concordium_event_schema_") {
+            if let Some(contract_name) = rest.strip_prefix("init_") {
+                let function_schema_event = generate_schema_run(&artifact, name.as_ref())?;
+
+                let contract_schema = contract_schemas
+                    .entry(contract_name.to_owned())
+                    .or_insert_with(schema::ContractV3::default);
+                contract_schema.event = function_schema_event;
```

Comment:
I'm a bit confused why this typechecks. Isn't `contract_schema.event: Option<Type>`?

So I would have expected you need to wrap this in a Some, unless your `concordium_event_schema` function serializes a `Some` which I do not see that it does.

So I kind of expect
```rust
contract_schema.event = Some(function_schema_event);
```

Am I misunderstanding something?

Reply:
I think that event is already in the format `Some(event_schema)` from this part here:

https://github.com/Concordium/concordium-rust-smart-contracts/blob/0e1331af82938f02d489db87659f782f87fb7235/concordium-std-derive/src/lib.rs#L1075

Reply:
I see, you are serializing an Option<Type> there already. This is a bit different from init I think. An alternative would be
```rust

#[cfg(feature = "build-schema")]
fn contract_function_schema_tokens2(
    event_option: Option<syn::LitStr>,
    contract_name: String,
) -> syn::Result<proc_macro2::TokenStream> {
    let mut embed = false;

    let event_schema = format_ident!("event");
    let event_embed = if let Some(event_ty) = event_option {
        let ty = event_ty.parse::<syn::Type>()?;
        let schema_name = format!("concordium_event_schema_{}", contract_name);
        let schema_ident = format_ident!("concordium_event_schema_{}", contract_name);
        quote! {
             let #event_schema = Some(<#ty as schema::SchemaType>::get_type());
             let schema_bytes = concordium_std::to_bytes(&#event_schema);
             concordium_std::put_in_memory(&schema_bytes)
        }
    } else {
        Ok(proc_macro2::TokenStream::new())
    }
```

and then here you know that if the function exists, then the event schema will be there, so you would  wrap it into Some.

Reply:
updated this pull request and the `concordium-rust-smart-contracts` one.

Reply:
Great, thank you.